### PR TITLE
Correctly combine fine TS from TLU

### PR DESCRIPTION
**Pull request by jekroege**
This MR contains changes in order to correctly combine the fine timestamps from the TLU to get the precise timestamp with 780 ps binning, solving the issue with the overflow of the counter (side peaks in time residuals).

Tested for 2 trigger inputs.

**Plan:** take data with 3-4 trigger inputs at DESY in July 2020 and then test other combinations as well.

**WIP** because not yet cleaned/fully tested.

Please comment on current implementation.

### CODE REVIEW

**Comment by simonspa** [comment]
Make sure to validate the values obtained before shifting bits around. We just found one run where this gives us a floating point exception...

**Comment by simonspa** [comment]
cc @schuetzepaul

File: user/tlu/module/src/TluRawEvent2StdEventConverter.cc [hunk 1]
```diff
@@ -30,28 +31,94 @@ bool TluRawEvent2StdEventConverter::Converting(eudaq::EventSPC d1, eudaq::Standa
...
+      finets_vec.emplace_back(finets5 & 0xFF);
+  }
+  if(finets_vec.size() < 2) {
+      std::cout << "ERROR: received less than 2 trigger inputs" << std::endl;
+      return false;
+
+      // or throw exception:
+      throw eudaq::DataInvalid("Received less than 2 trigger inputs.");
+
+      // or is that even fine to have only one trigger input???
```
Comment (simonspa) [inline user/tlu/module/src/TluRawEvent2StdEventConverter.cc]: I don't thing this constitutes an error condition. Whether it is a sensible thing to use a single noisy PMT without coincidence as trigger is a different question, but one to be solved by the user during test beam - there is not much point in telling them after the fact that this was probably a mistake.

**Comment by simonspa** [comment]
Hm, not sure why I still get lots of changes displayed which should be in `caribou` already...

File: user/tlu/module/src/TluRawEvent2StdEventConverter.cc (hunk 1 above)
Comment (jekroege) [reply user/tlu/module/src/TluRawEvent2StdEventConverter.cc]: That's true, or other users might prefer to use some NIM logic or whatever and provide a "ready-to-go" trigger signal. I've removed it.

**Comment by jekroege** [comment]
> Hm, not sure why I still get lots of changes displayed which should be in `caribou` already...

I merged the latest caribou again, this looks fine now.

**Comment by jekroege** [comment]
> Make sure to validate the values obtained before shifting bits around. We just found one run where this gives us a floating point exception...

Which values do you mean? And can you give me more details on when the exception happened? Which configuration? 2 triggers?

File: user/timepix3/module/src/Timepix3Event2StdEventConverter.cc [hunk 2]
```diff
@@ -139,7 +139,8 @@ bool Timepix3RawEvent2StdEventConverter::Converting(eudaq::EventSPC ev, eudaq::S
         // Hence, if the current timestamp is more than 20us earlier than the previous timestamp, we can assume that
         // a 2nd T0 has occured. Take 500us with some safety margin.
         // This implies we cannot detect a 2nd T0 within the first 500us after the initial T0. But did this ever happen?
-      } else if (m_syncTime < m_syncTime_prev - (500 * 4096 * 40)) {
```
Comment (jekroege) [inline user/timepix3/module/src/Timepix3Event2StdEventConverter.cc]: unrelated change, remove

**Comment by simonspa** [comment]
> > Make sure to validate the values obtained before shifting bits around. We just found one run where this gives us a floating point exception...
> 
> Which values do you mean? And can you give me more details on when the exception happened? Which configuration? 2 triggers?

You are throwing any string you read from `d1->GetTag()` into `std::stoi` which throws an exception if it cannot convert to an integer. Furthermore you are shifting bits around with `triggersFired` without boundary checks (try shifting by 33 bits and you're be killed by the kernel). 

Also, is your averaging correct? You are calculating ((a+b)/2 + c)/3 etc as far as I can see.

**Comment by jekroege** [comment]
> > > Make sure to validate the values obtained before shifting bits around. We just found one run where this gives us a floating point exception...
> > 
> > 
> > Which values do you mean? And can you give me more details on when the exception happened? Which configuration? 2 triggers?
> 
> You are throwing any string you read from `d1->GetTag()` into `std::stoi` which throws an exception if it cannot convert to an integer. Furthermore you are shifting bits around with `triggersFired` without boundary checks (try shifting by 33 bits and you're be killed by the kernel).
That's a good point! I shall add a check here and throw an exception.
> 
> Also, is your averaging correct? You are calculating ((a+b)/2 + c)/3 etc as far as I can see.
I think the averaging is correct, there's a little "i" to calculate (i times the current average + the new value) / (i+1)

**Comment by simonspa** [comment]
> > I think the averaging is correct, there's a little "i" to calculate (i times the current average + the new value) / (i+1)

Still not very beautiful...

**Comment by jekroege** [comment]
Ok, I've added a `try/catch` logic for the `stoi()` now to make sure that I get valid values there and don't crash.
Next, I'll test a bit more with different runs with 2/3 triggers.

Regarding the "rolling average" logic: I know it's not beautiful but I couldn't think of any other solution because I constantly need to check for a possible overflow. I might ask @tvanat for an idea here as well.

File: user/tlu/module/src/TluRawEvent2StdEventConverter.cc [hunk 3]
```diff
@@ -29,41 +30,126 @@ bool TluRawEvent2StdEventConverter::Converting(eudaq::EventSPC d1, eudaq::Standa
...
-               / __builtin_popcount(triggersFired & 0x3F); // count "ones" in binary
-
-  auto ts = (d1->GetTimestampBegin() * 1000) & 0xFFFFFFFFFFFFFF00 + static_cast<uint64_t>(25. / 32. * finets);
+  // Should I read these in at m_first_time and store in static member variable or read in each time?
+  double tof_scint0 = conf->Get("tof_scint0", 0.); // in nanoseconds
+  double tof_scint1 = conf->Get("tof_scint1", 0.); // in nanoseconds
+  double tof_scint2 = conf->Get("tof_scint2", 0.); // in nanoseconds
+  double tof_scint3 = conf->Get("tof_scint3", 0.); // in nanoseconds
+  double tof_scint4 = conf->Get("tof_scint4", 0.); // in nanoseconds
+  double tof_scint5 = conf->Get("tof_scint5", 0.); // in nanoseconds
```
Comment (jekroege) [inline user/tlu/module/src/TluRawEvent2StdEventConverter.cc]: input parameters in ns (without units) or in 780ps bins: up for discussion, I'd still opt for ns

**Comment by jekroege** [comment]
This works very nicely now and I have no better idea for the "overflow" detection logic, which needs to be generic enough to handle any case of n-out-of-m triggers like "2 or 3 out of 3".

Please have a look and let me know.

File: user/tlu/module/src/TluRawEvent2StdEventConverter.cc (hunk 3 above)
Comment (jekroege) [reply user/tlu/module/src/TluRawEvent2StdEventConverter.cc]: Changed to `delay_scint<i>` in 781.25ps bins now, README updated accordingly.

Also tested 3 cases, see screenshot below:
1) all delays zero
2) delays = geometric time-of-flight
3) delays = measured offsets in fineTS0 vs. fineTS1 histograms

As expected, 3) gives the best results.
The only problem left is: I extracted the delay values from the histograms fineTS0 vs. fineTS1 etc. in corryvreckan, see [here](https://gitlab.cern.ch/jekroege/corryvreckan/-/blob/lab_analysis/src/modules/EventLoaderEUDAQ2/EventLoaderEUDAQ2.cpp#L99).
in the `EventLoaderEUDAQ2`. 
Since it's not generic at all, I'm lacking an idea how to make this available in the master. Maybe a new module...?

image

File: user/tlu/module/src/TluRawEvent2StdEventConverter.cc [hunk 4]
```diff
@@ -29,41 +30,124 @@ bool TluRawEvent2StdEventConverter::Converting(eudaq::EventSPC d1, eudaq::Standa
...
+      // --> Convert to double in nanoseconds?
+  }
+
+  /*
+  // This works well: using ONLY fineTS0 OR fineTS1:
+  auto finets_avg = finets0;
+  auto finets_avg = finets1;
+  */
+
+  /*
```
Comment (simonspa) [inline user/tlu/module/src/TluRawEvent2StdEventConverter.cc]: Remove commented code please

**Review by simonspa — CHANGES_REQUESTED** [review]

File: user/tlu/module/src/TluRawEvent2StdEventConverter.cc (hunk 4 above)
Comment (jekroege) [reply user/tlu/module/src/TluRawEvent2StdEventConverter.cc]: Done.

**Review by simonspa — APPROVED** [review]
